### DISCUSSION TITLE
adding-content-negotiation-to-minimal-apis-with-carter/

### DISCUSSION THREAD

**Discussion by andrewlock** [andrewlock/blog-comments#374] [Announcements]
# adding-content-negotiation-to-minimal-apis-with-carter/

In this post I show how to use the open source library Carter to add content negotiation to an ASP.NET Core minimal API app

https://andrewlock.net/adding-content-negotiation-to-minimal-apis-with-carter/

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at July 13 2022, 11:15, in reply to [Graham King](https://disqus.com/by/BlackstarSolar/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/374#discussioncomment-6063154)</em>

---

<p>Good point, thanks, I always forget that, will update 🙂</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at July 13 2022, 11:15, in reply to [Muhammad Rehan Saeed](https://disqus.com/by/RehanSaeedUK/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/374#discussioncomment-6063159)</em>

---

<p>I'm interested what other "basic features" you feel are missing? With filters and route groups in .NET 7, and with conneg, validation helpers, and easy modularisation from Carter, I'm not sure if Web API gives a huge amount over minimal APIs? I don't see big gaps, so interested to know where my blind spots are 🙂 I guess simple replacement of System.Text.Json with <a href="http://json.net" rel="nofollow noopener" title="json.net">json.net</a> is one?</p><p>RE migrating - that's always a harder proposition, but for green field, I don't see any obvious issues?</p>

[a comment by an automated account is not shown]

**Comment by JohnGalt1717** [comment]
This just looks like Standing on your head to use minimal apis while reinventing controllers...

**Comment by TrayanZapryanov** [comment]
If you want to reuse your already defined MVC output formatters - here it is a way to do it using Carter.
Just build a custom IResponseNegotiator like described in the article, but delegate writing to your MVC formatter.
Sampe with XmlFormatter

```
internal class XmlResponseNegotiator : IResponseNegotiator
{
	private static readonly XmlSerializerOutputFormatter OutputFormatter = new XmlSerializerOutputFormatter();
	public bool CanHandle(MediaTypeHeaderValue accept)
		=> OutputFormatter.SupportedMediaTypes.Contains(accept.MediaType.ToString());

	private readonly Func<Stream, Encoding, TextWriter> _writerFactory;

	public XmlResponseNegotiator(IHttpResponseStreamWriterFactory writerFactory)
	{
		_writerFactory = writerFactory.CreateWriter;
	}
	public Task Handle<T>(HttpRequest request, HttpResponse res, T model, CancellationToken cancellationToken)
	{
		return OutputFormatter.WriteResponseBodyAsync(
			new OutputFormatterWriteContext(request.HttpContext, _writerFactory, typeof(T), model), Encoding.UTF8);
	}
}
```
